src/Data/Type/Rec.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Data.Type.Rec
+  ( type Rec,
+    constMap,
+    setRec,
+    getRec,
+    pattern RCon,
+    RecT (RConT, RNil),
+    fieldMap,
+    constMapT,
+    fields,
+    type (.|),
+    HasSel (..),
+
+    -- * Re-exports
+    Ascribe,
+    type (#),
+    Name (Name),
+  )
+where
+
+import Control.Natural (type (~>))
+import Data.Functor.Identity (Identity (Identity, runIdentity))
+import Data.Kind (Constraint, Type)
+import GHC.TypeLits (Symbol)
+
+import Data.Ascript (Ascribe, AscriptName, AscriptType, type (#))
+import Data.Name (Name (Name))
+
+-- -------------------------------------------------------------------------------------------------
+
+type Rec :: [Ascribe Symbol Type] -> Type
+type Rec ctx = RecT Identity ctx
+
+constMap :: Monoid (t a) => (forall s x. Name s -> x -> t a) -> Rec ctx -> t a
+constMap f = constMapT (\name -> f name . Identity)
+{-# INLINE constMap #-}
+
+setRec :: forall ctx s a. s # a .| ctx => Name s -> a -> Rec ctx -> Rec ctx
+setRec name x = setRecT name (Identity x)
+{-# INLINE setRec #-}
+
+getRec :: forall ctx s a. s # a .| ctx => Name s -> Rec ctx -> a
+getRec n r = runIdentity (getRecT n r)
+{-# INLINE getRec #-}
+
+pattern RCon :: () => (s # a ': ctx) ~ ctx' => Name s -> a -> Rec ctx -> Rec ctx'
```

Comment:
Any reason this couldn't be written without the type equality constraint, similar to how you're doing it with `RConT` on line 58?

```suggestion
pattern RCon :: () => Name s -> a -> Rec ctx -> Rec (s # a ': ctx)
```

And what is the empty constraint tuple / unit doing here?

Reply:
Bidirectional patterns have two contexts for whether when theyre used to pattern match or in an expression. The unit constraint in the first context is just saying we don't want to constrain line 53 and is also needed so I can constrain the destructive case. The type equality constraint in the second context tells GHC that this pattern matches for any `Rec ctx` that can be  substituted for `Rec (s # a ': ctx)`.

Reply:
I think you're suggesting these mean subtly different things?

```
f :: (C1 a, C2 b) => a -> b
f :: C1 a => C2 b => a -> b
```

If that's true... TIL 🙂 What are the differences? What does it mean to have "multiple contexts"?

Reply:
Yes, they're different in pattern synonyms but not here since f and g are functions:
```haskell
-- not sure if f is desugared into g or vice versa in GHC but these definitions are equivalent.
f :: (C1 a, C2 b) => a -> b
g :: C1 a => C2 b => a -> b 
```
There are different kinds of constraint contexts though. For functions, we can only talk about the required (aka needed) context, which says we need some `a` which satisfies `C1` (and `b` for `C2`) to be able to use `f` or `g`. However, for a bidirectional pattern synonym like
```haskell
Pat :: C1 a => C2 b => a -> T b
```
We start having different kinds of constraint contexts. The first one (the forward direction) is the same as it is for functions, we need `C1 a` to be able to use `Pat` on the RHS of an equation. The second context is the reverse. Here `C2 b` is in the "provided" context, saying that if we use `Pat` on the LHS of an equation we must introduce the constraint `C2 b`, which is why the `RCon` synonym works. Just as pattern matching on `RConT` for some `Rec ctx` would imply the constraint `ctx ~ s # a ': ctx'` given the type for the constructor `RConT`, the pattern `RCon` provides this same constraint whenever it is used on the LHS of an equation.

Reply:
Ah that makes sense; thank you for explaining!